stratum/hal/lib/tdi/es2k/es2k_sde_packetio.cc

```diff
@@ -0,0 +1,428 @@
+#include "stratum/hal/lib/tdi/tdi_sde_helpers.h"
+#include "stratum/lib/utils.h"
+#include "tdi/common/tdi_notifications.hpp"
+#include "tdi_rt/tdi_rt_defs.h"
+#include "tdi_rt/tdi_rt_operations.hpp"
+
+#define PKTIO_TABLE_NAME "pktio.packetio"
+#define RX_BURST_FIELD_ID 3
+
+enum PktIoField {
+  PORT_ID = 1,
+  QUEUE_ID = 2,
+  NB_PKTS = 3,
+  PKT_LEN = 4,
+  PKT_DATA = 5,
+  BURST_SZ = 6
+};
+
+enum NotificationId { RX = 1, TX = 2 };
+
+namespace stratum {
+namespace hal {
+namespace tdi {
+
+using namespace stratum::hal::tdi::helpers;
+
+// satish: Only for testing. Remove after testing
+void dump_pkt(const char* raw_data) {
+  struct ip* ip_hdr;
+  char src_ip_str[INET_ADDRSTRLEN];
+  char dst_ip_str[INET_ADDRSTRLEN];
+
+  ip_hdr = (struct ip*)(raw_data + 14);
+
+  // Extract and print the source and destination MAC and IP addresses
+  printf("pktio: Src MAC: ");
+  for (int j = 6; j < 12; ++j) {
+    printf("%02X", raw_data[j]);
+    if (j < 11) printf(":");
+  }
+  printf("\n");
+
+  printf("pktio: Dst MAC: ");
+  for (int j = 0; j < 6; ++j) {
+    printf("%02X", raw_data[j]);
+    if (j < 5) printf(":");
+  }
+  printf("\n");
+
+  inet_ntop(AF_INET, &(ip_hdr->ip_src), src_ip_str, INET_ADDRSTRLEN);
+  inet_ntop(AF_INET, &(ip_hdr->ip_dst), dst_ip_str, INET_ADDRSTRLEN);
+
+  printf("pktio: Src IP: %s\n", src_ip_str);
+  printf("pktio: Dst IP: %s\n", dst_ip_str);
+}
+
+::util::Status Es2kSdeWrapper::RegisterPacketReceiveWriter(
+    int device, std::unique_ptr<ChannelWriter<std::string>> writer) {
+  absl::WriterMutexLock l(&packet_rx_callback_lock_);
+  device_to_packet_rx_writer_[device] = std::move(writer);
+  return ::util::OkStatus();
+}
+
+::util::Status Es2kSdeWrapper::UnregisterPacketReceiveWriter(int device) {
+  absl::WriterMutexLock l(&packet_rx_callback_lock_);
+  device_to_packet_rx_writer_.erase(device);
+  return ::util::OkStatus();
+}
+
+::util::Status Es2kSdeWrapper::SetPacketIoConfig(
+    const PacketIoConfig& pktio_config) {
+  pktio_config_.CopyFrom(pktio_config);
+  return ::util::OkStatus();
+}
+
+// TxCallback that is called by sde for Tx done.
+void Es2kSdeWrapper::PktIoTxCallback(
+    std::unique_ptr<::tdi::TableKey> key,
+    std::unique_ptr<::tdi::TableData> data,
+    std::unique_ptr<::tdi::NotificationParams> pkts_info_recv, void* cookie) {
+  uint64_t port_id;
+  uint64_t queue_id;
+  uint64_t nb_pkts;
+  std::vector<uint64_t> pkt_len;  /* len of pkts */
+  std::vector<uint64_t> pkt_data; /* pointers to pkts data */
+
+  // parse callback params
+  pkts_info_recv->getValue(PORT_ID, &port_id);
+  pkts_info_recv->getValue(QUEUE_ID, &queue_id);
+  pkts_info_recv->getValue(NB_PKTS, &nb_pkts);
+  pkts_info_recv->getValue(PKT_LEN, &pkt_len);
+  pkts_info_recv->getValue(PKT_DATA, &pkt_data);
+
+  printf(
+      "[satish] pktio: callbacks received for %lu pkts sent on port_id: %lu, "
+      "queue_id: %lu\n",
+      nb_pkts, port_id, queue_id);
+
+  // freeing the buffers that were allocated during Tx
```

Comment:
```suggestion
  // Free the buffers now that transmission is complete
```

Not strictly necessary if your head-of-function comment says that the purpose of our callback function is to free the buffers, especially if you pare away the debug code and the code and variables that fetch parameters you don't use.

Reply:
Done.